3.assess_generalizability/scripts/generalizabilty_eval.py

```diff
@@ -0,0 +1,324 @@
+    )
+)
+
+# Remove rows where Metadata_genotype is "HET"
+plate6_df = plate6_df[plate6_df["Metadata_genotype"] != "HET"].reset_index(drop=True)
+
+# Count rows before dropping NaNs
+initial_count = plate6_df.shape[0]
+
+# Drop rows with NaNs
+plate6_df = plate6_df.dropna()
+
+# Count rows after dropping NaNs
+final_count = plate6_df.shape[0]
+
+# Print the count of dropped rows
+print(f"Dropped rows: {initial_count - final_count}")
+
+# Print shape and head of data
+print(plate6_df.shape)
+plate6_df.head()
+
+
+# ## Generate a shuffled dataset from the loaded in plate
+
+# In[5]:
+
+
+# Shuffle the features randomly, excluding columns that start with "Metadata_"
+shuffled_plate6_df = plate6_df.apply(
+    lambda x: rng.permutation(x) if not x.name.startswith("Metadata_") else x
+)
+
+# Print shape and head of data
+print(shuffled_plate6_df.shape)
+shuffled_plate6_df.head()
+
+
+# ## Apply model to final and shuffled versions of the plate data
+
+# In[6]:
+
+
+# Create list of the metadata columns only
+meta_cols = [col for col in plate6_df.columns if "Metadata" in col]
+
+# Define a dictionary to handle both data types
+data_dict = {"final": plate6_df, "shuffled": shuffled_plate6_df}
+
+# Initialize a list to store processed dataframes
+processed_dfs = []
+
+# Loop through the data dictionary to create probability dataframes
+for data_type, data in data_dict.items():
+    # Drop rows with Metadata_genotype == "HET"
+    data = data[data["Metadata_genotype"] != "HET"]
+
+    # Ensure no duplicates in data and reset index
+    data = data.drop_duplicates().reset_index(drop=True)
+
+    # Predict probabilities and labels
+    probabilities = model.predict_proba(data[model.feature_names_in_])[:, 1]
+    predicted_genotype = model.predict(
+        data[model.feature_names_in_]
+    )  # outputs as binary labels
+
+    # Convert true labels to binary values (0 or 1) for WT and Null
+    true_genotype = le.transform(data["Metadata_genotype"]).tolist()
+
+    # Create a dataframe with probabilities and predictions
+    probability_df = pd.DataFrame(
+        {
+            "probability_WT": probabilities,
+            "predicted_genotype": predicted_genotype,
+            "true_genotype": true_genotype,
+            "data_type": data_type,
+        },
+        index=data.index,  # Ensure alignment with original data
+    )
+
+    # Add metadata columns (reset index to align lengths)
+    metadata_df = data[meta_cols].reset_index(drop=True)
+    assert len(probability_df) == len(
+        metadata_df
+    ), "Row count mismatch between probabilities and metadata!"
+
+    full_df = pd.concat([probability_df, metadata_df], axis=1)
+    processed_dfs.append(full_df)
+
+# Combine all dataframes
+combined_df = pd.concat(processed_dfs, axis=0).reset_index(drop=True)
+
+# Save to Parquet (uncomment when needed)
+# combined_df.to_parquet(f"{model_dir}/plate_6_single_cell_probabilities.parquet")
```

Comment:
why commented out?

Reply:
Hmm I missed this, we _might_ want to do something with the probabilities later so I will save the data.